src/protocolsupport/protocol/packet/middle/clientbound/play/MiddleDeclareRecipes.java

```diff
@@ -1,22 +1,226 @@
 package protocolsupport.protocol.packet.middle.clientbound.play;
 
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
+
 import io.netty.buffer.ByteBuf;
+import protocolsupport.api.ProtocolVersion;
 import protocolsupport.protocol.ConnectionImpl;
 import protocolsupport.protocol.packet.middle.ClientBoundMiddlePacket;
+import protocolsupport.protocol.serializer.ItemStackSerializer;
 import protocolsupport.protocol.serializer.MiscSerializer;
+import protocolsupport.protocol.serializer.StringSerializer;
+import protocolsupport.protocol.serializer.VarNumberSerializer;
+import protocolsupport.protocol.utils.ProtocolVersionsHelper;
+import protocolsupport.protocol.utils.i18n.I18NData;
+import protocolsupport.protocol.utils.types.NetworkItemStack;
 
 public abstract class MiddleDeclareRecipes extends ClientBoundMiddlePacket {
 
 	public MiddleDeclareRecipes(ConnectionImpl connection) {
 		super(connection);
 	}
 
-	//TODO: structure
-	protected ByteBuf data;
+	protected static final Map<ProtocolVersion, ByteBuf> cachedBuffers = Collections.synchronizedMap(new EnumMap<>(ProtocolVersion.class));
+
+	protected Recipe[] recipes;
 
 	@Override
 	public void readFromServerData(ByteBuf serverdata) {
-		data = MiscSerializer.readAllBytesSlice(serverdata);
+		if (cachedBuffers.containsKey(connection.getVersion())) {
```

Comment:
This skips sending recipes entirely. Will break in case different players should have different recipes declared.

Reply:
Is it possible to have different recipes for different players in any current Bukkit implementation? It is some performance hit when not caching this, so I would like to have it there.

Reply:
Of course there is some performance hit. However this isn't a packet that is send alot. In general we read and write all packets actually, because we even want to support other plugins hacking with packets.

Reply:
Ok, I removed the cache